src/serac/physics/utilities/variational_form/integral.hpp

```diff
@@ -0,0 +1,591 @@
+   * @tparam spatial_dim The full dimension of the mesh
+   * @param[in] num_elements The number of elements in the mesh
+   * @param[in] J The Jacobians of the element transformations at all quadrature points
+   * @param[in] X The actual (not reference) coordinates of all quadrature points
+   * @see mfem::GeometricFactors
+   * @param[in] qf The user-provided quadrature function
+   * @note The @p Dimension parameters are used to assist in the deduction of the @a geometry_dim
+   * and @a spatial_dim template parameters
+   */
+  template <int geometry_dim, int spatial_dim, typename lambda_type>
+  Integral(int num_elements, const mfem::Vector& J, const mfem::Vector& X, Dimension<geometry_dim>,
+           Dimension<spatial_dim>, lambda_type&& qf)
+      : J_(J), X_(X)
+  {
+    constexpr auto geometry = supported_geometries[geometry_dim];
+    constexpr auto Q        = std::max(test_space::order, trial_space::order) + 1;
+
+    // these lines of code figure out the argument types that will be passed
+    // into the quadrature function in the finite element kernel.
+    //
+    // we use them to observe the output type and allocate memory to store
+    // the derivative information at each quadrature point
+    using x_t             = tensor<double, spatial_dim>;
+    using u_du_t          = typename lambda_argument<trial_space, geometry_dim, spatial_dim>::type;
+    using derivative_type = decltype(get_gradient(qf(x_t{}, make_dual(u_du_t{}))));
+
+    auto num_quadrature_points = static_cast<uint32_t>(X.Size() / spatial_dim);
+    qf_derivatives_.resize(sizeof(derivative_type) * num_quadrature_points);
+
+    // FIXME: Strict aliasing rule?
+    auto qf_derivatives_ptr = reinterpret_cast<derivative_type*>(qf_derivatives_.data());
+
+    // this is where we actually specialize the finite element kernel templates with
+    // our specific requirements (element type, test/trial spaces, quadrature rule, q-function, etc).
+    //
+    // std::function's type erasure lets us wrap those specific details inside a function with known signature
+    //
+    // note: the qf_derivatives_ptr is copied by value to each lambda function below,
+    //       to allow the evaluation kernel to pass derivative values to the gradient kernel
+    evaluation_ = [=](const mfem::Vector& U, mfem::Vector& R) {
+      evaluation_kernel<geometry, test_space, trial_space, geometry_dim, spatial_dim, Q>(U, R, qf_derivatives_ptr, J_,
+                                                                                         X_, num_elements, qf);
+    };
+
+    gradient_ = [=](const mfem::Vector& dU, mfem::Vector& dR) {
+      gradient_kernel<geometry, test_space, trial_space, geometry_dim, spatial_dim, Q>(dU, dR, qf_derivatives_ptr, J_,
+                                                                                       num_elements);
+    };
+  }
+
+  /**
+   * @brief Applies the integral, i.e., @a output_E = evaluate( @a input_E )
```

Comment:
I think it would be good to clarify mathematically how this integral is different from the gradient (maybe something like `integral of dot(shape_func_E,  evaluate( @2 input E)) dV`